### DISCUSSION TITLE
Installation of 4.2.4 failing - hiredis

### DISCUSSION THREAD

**Discussion by dhutchin-ctl** [irrdnet/irrd#641] [Q&A] [answered]
While trying to install IRRd v4.2.4, I'm running into a failure for hiredis.  Here are my steps:
1) Install pyenv
     unset PIP_REQUIRE_VERTUALENV
     unset PIPENV_VENV_IN_PROJECT
     curl https://pyenv.run | bash # Creates ~/.pyenv

2) update your PATH environment variable to include .pyenv/bin
     export PATH=$HOME/.pyenv/bin:$PATH

3) Initialize pyenv
     eval "$(pyenv init --path)"
     eval "$(pyenv init -)"
     eval "$(pyenv virtualenv-init -)"

4) Install a newer python version.
     pyenv install --list | less
     pyenv install 3.6.15 # if you want 3.6.15
     pyenv install 3.9.6  # if you want 3.9.6
     pyenv install pypy3.9-7.3.9 # (this is pypy7.3.9 which is python 3.9.12)

     ## Pick the version of python you want to use... so one of the follwowing
     pyenv global 3.6.15
     pyenv global 3.9.6
     pyenv global pypy3.9-7.3.9

     ## Now verify what you have
     which python
     python --version
     
5) Update pip
     python -m pip install --upgrade pip # get the latest version of pip so that warning messages go away

6) Install pipenv and pylint, these will go in ~/.local
     pip3 install --user pipenv pylint

7) update your PATH environment vairable to include .local/bin
     export PATH=$PATH:$HOME/.local/bin

8) set several environment variables that help enforce good habits
     export PIP_REQUIRE_VIRTUALENV=true
     export PYTHONDONTWRITEBYTECODE=1
     export PIPENV_VENV_IN_PROJECT=1

9) Build the specific project oriented virual environment
      pipenv install
        or, ignoring Pipfile:
      pipenv install irrd
    

But this is failing.  I can simplify the failure to simply:
    pipenv install hiredis

[irrd@rr2-dev ~]$ pipenv install hiredis
Installing hiredis...
Error:  An error occurred while installing hiredis!
Error text: Collecting hiredis
  Using cached hiredis-2.0.0.tar.gz (75 kB)
  Preparing metadata (setup.py): started
  Preparing metadata (setup.py): finished with status 'done'
Building wheels for collected packages: hiredis
  Building wheel for hiredis (setup.py): started
  Building wheel for hiredis (setup.py): finished with status 'error'
  Running setup.py clean for hiredis
Failed to build hiredis
Installing collected packages: hiredis
  Running setup.py install for hiredis: started
  Running setup.py install for hiredis: finished with status 'error'
  error: subprocess-exited-with-error
  × python setup.py bdist_wheel did not run successfully.
  │ exit code: 1
  ╰─> [32 lines of output]
      /tmp/pip-install-fm7wz9e3/hiredis_e9138c9b126e49bbb42fef8ec5645ddf/setup.py:7: DeprecationWarning: the imp module is deprecated in favour of importlib; see the module's documentation for alternative uses
        import sys, imp, os, glob, io
      /export/home/irrd/.venv/lib/pypy3.9/site-packages/setuptools/dist.py:757: UserWarning: Usage of dash-separated 'description-file' will not be supported in future versions. Please use the underscore name 'description_file' instead
        warnings.warn(
      running bdist_wheel
      running build
      running build_py
      creating build
      creating build/lib.linux-x86_64-3.9
      creating build/lib.linux-x86_64-3.9/hiredis
      copying hiredis/__init__.py -> build/lib.linux-x86_64-3.9/hiredis
      copying hiredis/version.py -> build/lib.linux-x86_64-3.9/hiredis
      copying hiredis/hiredis.pyi -> build/lib.linux-x86_64-3.9/hiredis
      copying hiredis/py.typed -> build/lib.linux-x86_64-3.9/hiredis
      warning: build_py: byte-compiling is disabled, skipping.
      running build_ext
      building 'hiredis.hiredis' extension
      creating build/temp.linux-x86_64-3.9
      creating build/temp.linux-x86_64-3.9/src
      creating build/temp.linux-x86_64-3.9/vendor
      creating build/temp.linux-x86_64-3.9/vendor/hiredis
      gcc -pthread -DNDEBUG -O2 -fPIC -Ivendor -I/export/home/irrd/.venv/include -I/export/home/irrd/.pyenv/versions/pypy3.9-7.3.9/include/pypy3.9 -c src/hiredis.c -o build/temp.linux-x86_64-3.9/src/hiredis.o
      gcc -pthread -DNDEBUG -O2 -fPIC -Ivendor -I/export/home/irrd/.venv/include -I/export/home/irrd/.pyenv/versions/pypy3.9-7.3.9/include/pypy3.9 -c src/reader.c -o build/temp.linux-x86_64-3.9/src/reader.o
      gcc -pthread -DNDEBUG -O2 -fPIC -Ivendor -I/export/home/irrd/.venv/include -I/export/home/irrd/.pyenv/versions/pypy3.9-7.3.9/include/pypy3.9 -c vendor/hiredis/alloc.c -o build/temp.linux-x86_64-3.9/vendor/hiredis/alloc.o
      gcc -pthread -DNDEBUG -O2 -fPIC -Ivendor -I/export/home/irrd/.venv/include -I/export/home/irrd/.pyenv/versions/pypy3.9-7.3.9/include/pypy3.9 -c vendor/hiredis/read.c -o build/temp.linux-x86_64-3.9/vendor/hiredis/read.o
      vendor/hiredis/read.c: In function ‘redisReaderFree’:
      vendor/hiredis/read.c:646:9: error: ‘for’ loop initial declarations are only allowed in C99 mode
               for (int i = 0; i < r->tasks; i++) {
               ^
      vendor/hiredis/read.c:646:9: note: use option -std=c99 or -std=gnu99 to compile your code
      error: command '/bin/gcc' failed with exit code 1
      [end of output]
  note: This error originates from a subprocess, and is likely not a problem with pip.
  ERROR: Failed building wheel for hiredis
  error: subprocess-exited-with-error
  × Running setup.py install for hiredis did not run successfully.
  │ exit code: 1
  ╰─> [34 lines of output]
      /tmp/pip-install-fm7wz9e3/hiredis_e9138c9b126e49bbb42fef8ec5645ddf/setup.py:7: DeprecationWarning: the imp module is deprecated in favour of importlib; see the module's documentation for alternative uses
        import sys, imp, os, glob, io
      /export/home/irrd/.venv/lib/pypy3.9/site-packages/setuptools/dist.py:757: UserWarning: Usage of dash-separated 'description-file' will not be supported in future versions. Please use the underscore name 'description_file' instead
        warnings.warn(
      running install
      /export/home/irrd/.venv/lib/pypy3.9/site-packages/setuptools/command/install.py:34: SetuptoolsDeprecationWarning: setup.py install is deprecated. Use build and pip and other standards-based tools.
        warnings.warn(
      running build
      running build_py
      creating build
      creating build/lib.linux-x86_64-3.9
      creating build/lib.linux-x86_64-3.9/hiredis
      copying hiredis/__init__.py -> build/lib.linux-x86_64-3.9/hiredis
      copying hiredis/version.py -> build/lib.linux-x86_64-3.9/hiredis
      copying hiredis/hiredis.pyi -> build/lib.linux-x86_64-3.9/hiredis
      copying hiredis/py.typed -> build/lib.linux-x86_64-3.9/hiredis
      warning: build_py: byte-compiling is disabled, skipping.
      running build_ext
      building 'hiredis.hiredis' extension
      creating build/temp.linux-x86_64-3.9
      creating build/temp.linux-x86_64-3.9/src
      creating build/temp.linux-x86_64-3.9/vendor
      creating build/temp.linux-x86_64-3.9/vendor/hiredis
      gcc -pthread -DNDEBUG -O2 -fPIC -Ivendor -I/export/home/irrd/.venv/include -I/export/home/irrd/.pyenv/versions/pypy3.9-7.3.9/include/pypy3.9 -c src/hiredis.c -o build/temp.linux-x86_64-3.9/src/hiredis.o
      gcc -pthread -DNDEBUG -O2 -fPIC -Ivendor -I/export/home/irrd/.venv/include -I/export/home/irrd/.pyenv/versions/pypy3.9-7.3.9/include/pypy3.9 -c src/reader.c -o build/temp.linux-x86_64-3.9/src/reader.o
      gcc -pthread -DNDEBUG -O2 -fPIC -Ivendor -I/export/home/irrd/.venv/include -I/export/home/irrd/.pyenv/versions/pypy3.9-7.3.9/include/pypy3.9 -c vendor/hiredis/alloc.c -o build/temp.linux-x86_64-3.9/vendor/hiredis/alloc.o
      gcc -pthread -DNDEBUG -O2 -fPIC -Ivendor -I/export/home/irrd/.venv/include -I/export/home/irrd/.pyenv/versions/pypy3.9-7.3.9/include/pypy3.9 -c vendor/hiredis/read.c -o build/temp.linux-x86_64-3.9/vendor/hiredis/read.o
      vendor/hiredis/read.c: In function ‘redisReaderFree’:
      vendor/hiredis/read.c:646:9: error: ‘for’ loop initial declarations are only allowed in C99 mode
               for (int i = 0; i < r->tasks; i++) {
               ^
      vendor/hiredis/read.c:646:9: note: use option -std=c99 or -std=gnu99 to compile your code
      error: command '/bin/gcc' failed with exit code 1
      [end of output]
  note: This error originates from a subprocess, and is likely not a problem with pip.
error: legacy-install-failure
× Encountered error while trying to install package.
╰─> hiredis
note: This is an issue with the package mentioned above, not pip.
hint: See above for output from the failure.
✘ Installation Failed 
[irrd@rr2-dev ~]$

**Comment by dhutchin-ctl** [comment, marked as answer]
I set CFLAGS in the environment to include -std=c99, and that fixed it.  Is that the right fix?

**Reply by mxsasha** [reply]
[It seems to be](https://stackoverflow.com/a/29338269), though it's curious that nobody has encountered this before. Hiredis has been a dependency since 4.1. It is possible this is specific to certain platforms due to hiredis having pre-compiled wheels for others, which skip local compilation and therefore would never encounter this error.

Which gcc version are you using?

**Comment by dhutchin-ctl** [comment]
$ gcc --version
gcc (GCC) 4.8.5 20150623 (Red Hat 4.8.5-44)



This is on a stock CentOS 7 box.  When I go to production, it will be RedHat 8
________________________________
